connor/advancedPoolModule.go

```diff
@@ -0,0 +1,228 @@
+package connor
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"math/big"
+	"time"
+
+	"github.com/sonm-io/core/connor/database"
+	"github.com/sonm-io/core/connor/watchers"
+	"github.com/sonm-io/core/proto"
+	"go.uber.org/zap"
+)
+
+//Tracking hashrate with using Connor's blacklist. Get data for 1 hour and another time => Detecting deviation.
+func (p *PoolModule) AdvancedPoolHashrateTracking(ctx context.Context, reportedPool watchers.PoolWatcher, avgPool watchers.PoolWatcher) error {
+	workers, err := p.c.db.GetWorkersFromDB()
+	if err != nil {
+		p.c.logger.Error("cannot get worker from pool DB", zap.Error(err))
+		return err
+	}
+
+	for _, w := range workers {
+		if w.BadGuy > 5 {
+			continue
+		}
+		if w.Iterations == 0 {
+			newIteration := w.Iterations + 1
+
+			if err := p.c.db.UpdateIterationPoolDB(newIteration, w.DealID); err != nil {
+				return err
+			}
+			continue
+		}
+
+		p.c.logger.Info("Update iteration worker less than 4", zap.Int64("worker", w.DealID))
+
+		newIteration := w.Iterations + 1
+		if err := p.c.db.UpdateIterationPoolDB(newIteration, w.DealID); err != nil {
+			return err
+		}
+
+		dealInfo, err := p.c.DealClient.Status(ctx, sonm.NewBigInt(big.NewInt(0).SetInt64(w.DealID)))
+		if err != nil {
+			log.Printf("Cannot get deal from market %v\r\n", w.DealID)
+			return err
+		}
+
+		bidHashrate, err := p.ReturnBidHashrateForDeal(ctx, dealInfo)
+		if err != nil {
+			return err
+		}
+
+		if w.Iterations < numberOfIterationsForH1 {
+			if err := p.UpdatePoolData(ctx, reportedPool, p.c.cfg.PoolAddress.EthPoolAddr, int64(PoolTypeReportedHashrate)); err != nil {
+				return err
+			}
+
+			if w.WorkerReportedHashrate == 0.0 {
```

Comment:
Comparing floats with zero is [tricky](https://bitbashing.io/comparing-floats.html).